shotover-proxy/src/frame/cassandra.rs

```diff
@@ -471,38 +503,497 @@ impl CassandraOperation {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Returns true if there are any parameters in the query
+    pub fn has_params(statement: &CassandraStatement) -> bool {
+        match statement {
+            CassandraStatement::Delete(delete) => {
+                if CQLStatement::has_params_in_relation_elements(&delete.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&delete.if_clause) {
+                    return true;
+                }
+            }
+            CassandraStatement::Insert(insert) => {
+                if let InsertValues::Values(operands) = &insert.values {
+                    for operand in operands {
+                        if let Operand::Param(_) = operand {
+                            return true;
+                        }
+                    }
+                }
             }
-            Ok(mut ast) => CQL::Parsed(Box::new(ast.remove(0))),
-            Err(err) => {
-                tracing::error!("Failed to parse CQL for frame {:?}\nError: {:?}", sql, err);
-                CQL::FailedToParse(sql)
+            CassandraStatement::Select(select) => {
+                return CQLStatement::has_params_in_relation_elements(&select.where_clause);
             }
+            CassandraStatement::Update(update) => {
+                for assignment_element in &update.assignments {
+                    if let Operand::Param(_) = &assignment_element.value {
+                        return true;
+                    }
+                    if let Some(assignment_operator) = &assignment_element.operator {
+                        match assignment_operator {
+                            AssignmentOperator::Plus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                            AssignmentOperator::Minus(operand) => {
+                                if let Operand::Param(_) = operand {
+                                    return true;
+                                }
+                            }
+                        }
+                    }
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.where_clause) {
+                    return true;
+                }
+                if CQLStatement::has_params_in_relation_elements(&update.if_clause) {
+                    return true;
+                }
+            }
+            _ => {}
+        }
+        false
+    }
+}
+
+impl Display for CQLStatement {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        self.statement.fmt(f)
+    }
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct CQL {
+    pub statements: Vec<CQLStatement>,
+    pub(crate) has_error: bool,
```

Comment:
Why do we have duplicate state via this has_error field?
Given that transforms can modify message state we could very easily end up with out of sync state.
Shouldnt the user just check for `CassandraStatement::Unknown`?

Reply:
Oh, best effort sounds wrong to me?
I think we would be better off completely failing to parse the message if we hit an error.
Better to pass through a message unmodified or drop it completely instead of modifying or redirecting it in a potentially dangerous way.

Reply:
I know this will mean some reworking of the parser and the PR but I think its really valuable.

Something like the existing:
```rust
pub enum CQL {
    // Box is used because Statement is very large
    Parsed(Box<Statement>),
    FailedToParse(String),
}
```
Is how I would go about it but let me know if you need more concrete examples of how this would work.
And feel free to bring this up in slack to get more opinions.